turtlebot3_fake_node/src/turtlebot3_fake_node.cpp

```diff
@@ -0,0 +1,266 @@
+  this->get_parameter_or<std::string>("odom_frame", odom_.header.frame_id, "odom");
+  this->get_parameter_or<std::string>("base_frame", odom_.child_frame_id, "base_footprint");
+  this->get_parameter_or<double>("wheels.separation", wheel_seperation_, 0.0);
+  this->get_parameter_or<double>("wheels.radius", wheel_radius_, 0.0);
+}
+
+void Turtlebot3Fake::init_variables()
+{
+  // Initialise variables
+  wheel_speed_cmd_[LEFT]  = 0.0;
+  wheel_speed_cmd_[RIGHT] = 0.0;
+  goal_linear_velocity_   = 0.0;
+  goal_angular_velocity_  = 0.0;
+  cmd_vel_timeout_        = 1.0;
+  last_position_[LEFT]    = 0.0;
+  last_position_[RIGHT]   = 0.0;
+  last_velocity_[LEFT]    = 0.0;
+  last_velocity_[RIGHT]   = 0.0;
+
+  // TODO: Find more accurate covariance
+  // double pcov[36] = { 0.1,   0,   0,   0,   0, 0,
+  //                       0, 0.1,   0,   0,   0, 0,
+  //                       0,   0, 1e6,   0,   0, 0,
+  //                       0,   0,   0, 1e6,   0, 0,
+  //                       0,   0,   0,   0, 1e6, 0,
+  //                       0,   0,   0,   0,   0, 0.2};
+  // memcpy(&(odom_.pose.covariance), pcov, sizeof(double)*36);
+  // memcpy(&(odom_.twist.covariance), pcov, sizeof(double)*36);
+
+  odom_pose_[0] = 0.0;
+  odom_pose_[1] = 0.0;
+  odom_pose_[2] = 0.0;
+  odom_vel_[0] = 0.0;
+  odom_vel_[1] = 0.0;
+  odom_vel_[2] = 0.0;
+
+  joint_states_.name.push_back("wheel_left_joint");
+  joint_states_.name.push_back("wheel_right_joint");
+  joint_states_.position.resize(2, 0.0);
+  joint_states_.velocity.resize(2, 0.0);
+  joint_states_.effort.resize(2, 0.0);
+
+  rclcpp::Clock clock(RCL_SYSTEM_TIME);
+  prev_update_time_ = clock.now();
+}
+
+/********************************************************************************
+** Callback functions for ROS subscribers
+********************************************************************************/
+void Turtlebot3Fake::command_velocity_callback(const geometry_msgs::msg::Twist::SharedPtr cmd_vel_msg)
+{
+  rclcpp::Clock clock(RCL_SYSTEM_TIME);
+  last_cmd_vel_time_ = clock.now();
+
+  goal_linear_velocity_  = cmd_vel_msg->linear.x;
+  goal_angular_velocity_ = cmd_vel_msg->angular.z;
+
+  wheel_speed_cmd_[LEFT]  = goal_linear_velocity_ - (goal_angular_velocity_ * wheel_seperation_ / 2);
+  wheel_speed_cmd_[RIGHT] = goal_linear_velocity_ + (goal_angular_velocity_ * wheel_seperation_ / 2);
+}
+
+/********************************************************************************
+** Update functions
+********************************************************************************/
+void Turtlebot3Fake::update_callback()
+{
+  rclcpp::Clock clock(RCL_SYSTEM_TIME);
+  rclcpp::Time time_now = clock.now();
+  rclcpp::Duration duration = time_now - prev_update_time_;
+  prev_update_time_ = time_now;
+
+  // zero-ing after timeout (stop the robot if no cmd_vel)
+  if ((time_now - last_cmd_vel_time_).seconds() > cmd_vel_timeout_)
```

Comment:
How about to change seconds() to nanoseconds() to frequently make control input to robot

Reply:
Let's use smaller units.

Reply:
@routiful @robotpilot 
I changed the mentioned lines as [this](https://github.com/ROBOTIS-GIT/turtlebot3_simulations/pull/90/commits/4fa9c03d8c7bb3ae38fdfbc95a6229c8db190e87) 
Please change the same lines in the function calculate_odometry in the [file](https://github.com/ROBOTIS-GIT/turtlebot3/blob/ros2/turtlebot3_node/src/odometry.cpp) in the next update.